File: L03/codes/L03_noise.py
# -*- coding: utf-8 -*-

import numpy as np
import matplotlib.pyplot as plt

from scipy.stats import norm
from scipy.stats import lognorm

#------------------------------------------------------------------------------
# Function, main()
#------------------------------------------------------------------------------
def L03_main():
    
    # generate data for specific probability density function (PDF)
    N = 10000
    x0 = np.random.randn(N)
    x1 = np.random.lognormal(0.5, 0.5, size=N)
    
    # plot
    plt.subplot(2, 2, 1) 
    ax = plt.plot(x0)
    plt.setp(ax, color='k', linewidth=0.5)
    plt.xlim(0, N)
    
    plt.subplot(2, 2, 3)
    ax = plt.plot(x1)
    plt.setp(ax, color='k', linewidth=0.5)
    plt.xlim(0, N)
    
    # histogram - normal
    xmin = np.min(x0) 
    xmax = np.max(x0)
    bx = np.linspace(xmin, xmax, 100)
    hx, bx = np.histogram(x0, bx)
    hx = hx / np.sum(hx)
    bx = bx[:-1] + (bx[1] - bx[0]) / 2 # get center of bins
    
    # fit histogram
    mu, std = norm.fit(x0)
    px = norm.pdf(bx, mu, std)
    px = px / np.sum(px)
    
    # plot histogram
    plt.subplot(2, 2, 2) 
    plt.bar(bx, hx, color='k', width=0.05)
    plt.plot(bx, px, 'r')
    
    # histogram - lognormal
    xmin = np.min(x1) 
    xmax = np.max(x1)
    bx = np.linspace(xmin, xmax, 100)
    hx, bx = np.histogram(x1, bx)
    hx = hx / np.sum(hx)
    bx = bx[:-1] + (bx[1] - bx[0]) / 2 # get center of bins
    
    # fit histogram
    p0, p1, p2 = lognorm.fit(x1)
    px = lognorm.pdf(bx, p0, p1, p2)
    px = px / np.sum(px)
    
    # plot histogram
    plt.subplot(2, 2, 4) 
    plt.bar(bx, hx, color='k', width=0.05)
    plt.plot(bx, px, 'r')
    
    # set layout
    plt.tight_layout(pad=0.5, w_pad=0.5, h_pad=1.0)

#------------------------------------------------------------------------------
if __name__ == '__main__':
    L03_main()

File: L03/figures/L03_noise.m
%-------------------------------------------------------------------------------
% Function
%-------------------------------------------------------------------------------
function L03_noise()

clc;

% generate distributions
N = 10000;
x0 = randn(N, 1);
x1 = lognrnd(0.5, 0.5, N, 1);

% histogram - normal
xmin = min(x0);
xmax = max(x0);
bx0 = linspace(xmin, xmax, 100);
hx0 = histc(x0, bx0);
hx0 = hx0 / sum(hx0);

% fit data - normal
[mu, sigma] = normfit(x0);
px0 = normpdf(bx0, mu, sigma);
px0 = px0 / sum(px0);

% histogram - lognormal
xmin = min(x1);
xmax = max(x1);
bx1 = linspace(xmin, xmax, 100);
hx1 = histc(x1, bx1);
hx1 = hx1 / sum(hx1);

% fit data - lognormal
p = lognfit(x1);
px1 = lognpdf(bx1, p(1), p(2));
px1 = px1 / sum(px1);

% * figure 1
figure(1);

% plot
subplot(2, 1, 1);
plot(x0, 'k');
xlim([1, N]);
set(gca, 'FontSize', 12);
set(get(gca, 'XLabel'), 'String', 'samples', 'FontSize', 12);
set(get(gca, 'YLabel'), 'String', 'amplitude', 'FontSize', 12);

% plot histogram
subplot(2, 1, 2);
bar(bx0, hx0, 'k'); hold on;
plot(bx0, px0, 'r', 'LineWidth', 2);
xlim([-3.5, 3.5]);
set(gca, 'FontSize', 12);
set(get(gca, 'XLabel'), 'String', 'amplitude', 'FontSize', 12);
set(get(gca, 'YLabel'), 'String', 'counts', 'FontSize', 12);

% * figure 2
figure(2);

% plot
subplot(2, 1, 1);
plot(x1, 'k');
xlim([1, N]);
set(gca, 'FontSize', 12);
set(get(gca, 'XLabel'), 'String', 'samples', 'FontSize', 12);
set(get(gca, 'YLabel'), 'String', 'amplitude', 'FontSize', 12);

% plot histogram
subplot(2, 1, 2);
bar(bx1, hx1, 'k'); hold on;
plot(bx1, px1, 'r', 'LineWidth', 2);
xlim([0, 7]);
set(gca, 'FontSize', 12);
set(get(gca, 'XLabel'), 'String', 'amplitude', 'FontSize', 12);
set(get(gca, 'YLabel'), 'String', 'counts', 'FontSize', 12);

end % end

%-------------------------------------------------------------------------------
